cpp/src/bitmask/bit_mask.h

```diff
@@ -0,0 +1,291 @@
+   *  @param[in]  valid         The bit mask to update
+   *  @param[in]  record_idx    The record index
+   *
+   *  @return which bit within the bit mask
+   */
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE
+  bool is_valid(const bit_mask_t *valid, T record_idx) {
+    static_assert(std::is_integral<T>::value, "Record index must be of an integral type");
+
+    const gdf_size_type rec{details::which_element(record_idx)};
+    const gdf_size_type bit{details::which_bit(record_idx)};
+
+    return ((valid[rec] & (bit_mask_t{1} << bit)) != 0);
+  }
+
+  /**
+   *  @brief set the specified bit in the bit mask in an unsafe manner
+   *
+   *  This function sets the specified bit in an unsafe manner.  It assumes that
+   *  that the calling code guarantees a thread-safe context.  That is, either
+   *  the function is called from a block of serial code, or the data is distributed
+   *  among the threads such that no two threads could be updating a bit in the
+   *  same memory location concurrently.
+   *
+   *  @param[in,out]  valid         The bit mask to update
+   *  @param[in]      record_idx    The record index
+   *
+   */
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE
+  void set_bit_unsafe(bit_mask_t *valid, T record_idx) {
+    static_assert(std::is_integral<T>::value, "Record index must be of an integral type");
+
+    const gdf_size_type rec{details::which_element(record_idx)};
+    const gdf_size_type bit{details::which_bit(record_idx)};
+
+    valid[rec] = valid[rec] | (bit_mask_t{1} << bit);
+  }
+
+  /**
+   *  @brief clear the specified bit in the bit mask in an unsafe manner.
+   *
+   *  This function clears the specified bit in an unsafe manner.  It assumes that
+   *  that the calling code guarantees a thread-safe context.  That is, either
+   *  the function is called from a block of serial code, or the data is distributed
+   *  among the threads such that no two threads could be updating a bit in the
+   *  same memory location concurrently.
+   *
+   *  @param[in,out]  valid         The bit mask to update
+   *  @param[in]      record_idx    The record index
+   *
+   */
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE
+  void clear_bit_unsafe(bit_mask_t *valid, T record_idx) {
+    static_assert(std::is_integral<T>::value, "Record index must be of an integral type");
+
+    const gdf_size_type rec{details::which_element(record_idx)};
+    const gdf_size_type bit{details::which_bit(record_idx)};
+
+    valid[rec] = valid[rec] & (~(bit_mask_t{1} << bit));
+  }
+
+  /**
+   *  @brief set the specified bit in the bit mask in an threadsafe manner
+   *
+   *  This function sets the specified bit in an threadsafe manner.  It uses
+   *  atomic memory operations to guarantee that no update interferes with
+   *  on another.
+   *
+   *  @param[in,out]  valid         The bit mask to update
+   *  @param[in]      record_idx    The record index
+   *
+   */
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE
```

Comment:
I don't believe this function is HOST-callable, so I would change this to just __device__ and use a different implementation on the host. `atomicOr` is not a HOST-callable function.

Reply:
~@ChuckHastings I suggest wrapping the `atomicOr` in a `#ifdef __CUDACC__`, then provide an `#elif` host-side implementation. That way this function can be used both on the device and host (for testing purposes).~

~This goes for any bit-utility function that uses device specific code (e.g., atomics).~

Nevermind, the `*_unsafe` functions fulfill the host-side implementation. Make any function that is device only to be `CUDA_DEVICE_CALLABLE`.